### PR TITLE
Fix ReadAsync not returning when cancellation token fires during FlushAsync

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/dotnet/corefx/issues/31201

### CODE REVIEW

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 1]
```diff
@@ -336,6 +336,19 @@ internal ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken)
                 // AttachToken before completing reader awaiter in case cancellationToken is already completed
                 cancellationTokenRegistration = _writerAwaitable.AttachToken(cancellationToken, s_signalWriterAwaitable, this);
 
+                // If the writer is completed (which it will be most of the time) the return a completed ValueTask
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: nit: **then** return a completed ValueTask

File: src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs [hunk 2]
```diff
@@ -325,6 +325,81 @@ public void GetResultThrowsIfFlushAsyncTokenFiredAfterCancelPending()
...
+
+            var canceller = Task.Run(() =>
+            {
+                while (!writer.IsCompleted)
+                {
+                    cancellationTokenSource.Cancel();
+                }
+            });
+
+            Assert.True(Task.WaitAll(new [] { writer, reader, canceller }, TimeSpan.FromSeconds(30)), "Reader was not completed in reasonable time");
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs]: Isn't potentially spending 30 seconds on a single unit test too long? Does the test still work as intended if we change this to < 5 seconds?

File: src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs [hunk 3]
```diff
@@ -325,6 +325,81 @@ public void GetResultThrowsIfFlushAsyncTokenFiredAfterCancelPending()
...
+            // This test tries to get pipe into a state where ReadAsync is being awaited
+            // and FlushAsync is cancelled while the method is running
+            Pipe = new Pipe();
+            var autoResetEvent = new ManualResetEvent(false);
+
+            var cancellationTokenSource = new CancellationTokenSource();
+            var writer = Task.Run(async () =>
+            {
+                var cancellations = 0;
+                while (cancellations < 20)
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs]: Any particular reason for 20?

File: src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs [hunk 4]
```diff
@@ -325,6 +325,81 @@ public void GetResultThrowsIfFlushAsyncTokenFiredAfterCancelPending()
...
+                            cancellationTokenSource = new CancellationTokenSource();
+                            continue;
+                        }
+
+                        await Pipe.Writer.FlushAsync(cancellationTokenSource.Token);
+                        autoResetEvent.Reset();
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        cancellations ++;
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs]: nit: remove space

File: src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs (hunk 2 above)
Comment (pakrym) [reply src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs]: Only if fails. I don't want to introduce the test that would be flaky on slower machines. In the successful case, it passes quickly.

File: src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs (hunk 3 above)
Comment (pakrym) [reply src/System.IO.Pipelines/tests/FlushAsyncCancellationTests.cs]: No, just a random number.

**Pull request merged**